depinject/container.go

```diff
@@ -130,7 +124,40 @@ func (c *container) getResolver(typ reflect.Type, key string) (resolver, error)
 		c.resolvers[mapType] = &mapOfOnePerModuleResolver{r}
 	}
 
-	return c.resolvers[typ], nil
+	res := c.resolvers[typ]
+
+	if res == nil && typ.Kind() == reflect.Interface {
+		var matches []reflect.Type
+		for k := range c.resolvers {
+			if k.Kind() != reflect.Interface && k.Implements(typ) {
+				matches = append(matches, k)
+			}
+		}
+
+		if len(matches) == 1 {
+			res = c.resolvers[matches[0]]
+			c.logf("Implicitly registering resolver %v for interface type %v", matches[0], typ)
+		} else if len(matches) > 1 {
+			p, found := findPreference(c.preferences, typ, key)
```

Comment:
That is the intent as written [here](https://github.com/cosmos/cosmos-sdk/blob/0a586840c593e61d6aec0f40754d26c8f94b6e9b/depinject/container.go#L155-L157).  A hit on `c.resolvers[typ]` *should* skip the whole block, have I mucked it up or misunderstood?

Reply:
when a user registers a preference, there should be a map based on the in type name. when there's a request to resolve that in type, the preference should be used instead of any other search. it should supersede even a direct match